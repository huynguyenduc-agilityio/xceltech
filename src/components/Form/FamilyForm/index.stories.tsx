import type { Meta, StoryObj } from '@storybook/react';

// Components
import FamilyForm from '.';

const meta = {
  title: 'Components/FamilyForm',
  component: FamilyForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  render: () => (
    <div className="flex items-center justify-center w-screen py-20 bg-blue-light">
      <div className="bg-white w-[988px] rounded-[15px]">
        <div className="p-10">
          <FamilyForm />
        </div>
      </div>
    </div>
  ),
} satisfies Meta<typeof FamilyForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
};
