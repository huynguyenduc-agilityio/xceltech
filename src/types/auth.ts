export interface ILoginUser {
  email: string;
  password: string;
}

export interface IAuthUser {
  refresh: string;
  access: string;
  user: {
    id: string;
    email: string;
  };
}

export interface IRegisterUser extends ILoginUser {
  firstName: string;
  lastName: string;
  phone: string;
  confirmPassword: string;
  isAgreeTerms?: boolean;
  isReceiveNewsletters?: boolean;
}

export interface IActiveUser {
  uidb64: string | null;
  token: string | null;
}
