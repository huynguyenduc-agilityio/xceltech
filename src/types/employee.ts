import { z } from 'zod';

import {
  educationSchema,
  familySchema,
  financialSchema,
  guarantorSchema,
} from '@/utils';

export type EducationFormValues = z.infer<typeof educationSchema>;
export type GuarantorFormValues = z.infer<typeof guarantorSchema>;
export type FamilyFormValues = z.infer<typeof familySchema>;
export type FinancialFormValues = z.infer<typeof financialSchema>;

export interface IEmployeeEducationInfo extends EducationFormValues {
  id?: string;
  startDate?: Date | string;
  endDate?: Date | string;
  type?: EducationType;
}

export enum EducationType {
  Academic = 'Academic',
  Professional = 'Professional',
}

export interface IEmployeeGuarantor extends GuarantorFormValues {
  id?: string;
}

export interface IEmployeeFamilyInfo extends FamilyFormValues {
  id?: string;
}

export interface IEmployeeFinancialInfo extends FinancialFormValues {
  id?: string;
}

export type DocumentsForm = {
  documents: {
    offerLetter?: File | null;
    birthCertificate?: File | null;
    guarantorForm?: File | null;
    degreeCertificate?: File | null;
  };
};

export type Documents = {
  documentFile: string;
  documentType: string;
};
