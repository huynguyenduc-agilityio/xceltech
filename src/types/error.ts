export type ErrorType = {
  detail: string;
};
