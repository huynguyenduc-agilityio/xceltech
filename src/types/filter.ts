export interface FilterCriteria {
  [key: string]: string | number | boolean | string[];
}
