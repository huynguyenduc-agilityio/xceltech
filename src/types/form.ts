export interface UploadFileForm {
  [key: string]: File | null;
}

export type TField<T> = {
  name: keyof T;
  label: string;
  placeholder: string;
  min?: string;
  max?: string;
  rules: {
    required: string;
    pattern: {
      value: RegExp;
      message: string;
    };
  };
  ariaLabel: string;
};

export enum MutationType {
  Create = 'create',
  Edit = 'edit',
}
