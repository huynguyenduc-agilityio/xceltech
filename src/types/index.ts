export * from './status';
export * from './employee';
export * from './table';
export * from './auth';
export * from './form';
export * from './tab';
export * from './user';
export * from './queryKey';
export * from './leave';
export * from './filter';
export * from './error';
export * from './notification';
export * from './profile';
