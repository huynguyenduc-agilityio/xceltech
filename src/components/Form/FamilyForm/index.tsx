import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';

// Constants
import { MESSAGES } from '@/constants';

// Types
import {
  IEmployeeFamilyInfo,
  MutationType,
  ToastStatus,
  FamilyFormValues,
} from '@/types';

// Components
import {
  Button,
  Form,
  FormControl,
  FormField,
  FormItem,
  FormMessage,
  Label,
  Textarea,
  TextField,
} from '@/components';

// Utils
import { familySchema } from '@/utils/schemas/employee';

// Hooks
import { useFamilyMutation, useToast } from '@/hooks';

export interface IFamilyForm {
  mode: MutationType;
  initialValues?: Partial<IEmployeeFamilyInfo>;
  onBack?: () => void;
}

const FamilyForm = ({ initialValues = {}, mode, onBack }: IFamilyForm) => {
  const { toast } = useToast();
  const { handleFamilyMutation, isFamilyMutationLoading } = useFamilyMutation({
    type: mode,
  });

  const {
    fullName = '',
    relationship = '',
    phone = '',
    address = '',
  } = initialValues || {};
  const defaultValues: IEmployeeFamilyInfo = {
    fullName,
    relationship,
    phone,
    address,
  };

  const form = useForm<FamilyFormValues>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    resolver: zodResolver(familySchema),
    defaultValues,
  });

  const {
    control,
    handleSubmit,
    reset,
    getValues,
    formState: { isValid, isDirty, isSubmitting },
  } = form;

  const isCreate = mode === MutationType.Create;

  const onSubmit = async (data: IEmployeeFamilyInfo) => {
    const payload = {
      ...data,
      ...(!isCreate && initialValues?.id ? { id: initialValues.id } : {}),
    };

    try {
      await handleFamilyMutation(payload);

      toast({
        status: ToastStatus.Success,
        title: isCreate
          ? MESSAGES.COMMON.ADD_SUCCESS('Family')
          : MESSAGES.COMMON.UPDATE_SUCCESS('Family'),
      });
      onBack?.();

      reset(initialValues);
    } catch {
      toast({
        status: ToastStatus.Error,
        title: isCreate
          ? MESSAGES.COMMON.ADD_FAILED('Family')
          : MESSAGES.COMMON.UPDATE_FAILED('Family'),
      });

      reset(getValues());
    }
  };

  const disableSubmit = !isDirty || !isValid || isSubmitting;

  return (
    <Form {...form}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="w-full text-black-default space-y-10">
          <FormField
            control={control}
            name="fullName"
            render={({ field, fieldState: { error } }) => (
              <TextField
                label="Full Name"
                variant="primary"
                labelClassName="text-md"
                inputClassName="h-[68px] px-12 rounded-[15px]"
                placeholder="Enter your full name"
                {...field}
                errorMessage={error?.message}
              />
            )}
          />
          <div className="flex items-start gap-11 justify-between">
            <FormField
              control={control}
              name="relationship"
              render={({ field, fieldState: { error } }) => (
                <TextField
                  label="Relationship"
                  variant="primary"
                  className="space-y-5"
                  labelClassName="text-md"
                  inputClassName="h-[68px] px-12 rounded-[15px]"
                  placeholder="Enter your relationship"
                  {...field}
                  errorMessage={error?.message}
                />
              )}
            />
            <FormField
              control={control}
              name="phone"
              render={({ field, fieldState: { error } }) => (
                <TextField
                  label="Phone No"
                  variant="primary"
                  className="space-y-5"
                  labelClassName="text-md"
                  inputClassName="h-[68px] px-12 rounded-[15px]"
                  placeholder="Enter your phone number"
                  {...field}
                  errorMessage={error?.message}
                />
              )}
            />
          </div>
          <FormField
            control={control}
            name="address"
            render={({ field }) => (
              <FormItem>
                <Label className="text-md">Address</Label>
                <FormControl>
                  <Textarea
                    {...field}
                    placeholder="Enter your address"
                    className="px-11"
                  />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <div className="w-full h-[70px]">
            <Button
              type="submit"
              className="w-[364px] bg-green-primary"
              isLoading={isFamilyMutationLoading}
              disabled={disableSubmit}
            >
              {isCreate ? 'Submit' : 'Update'}
            </Button>
          </div>
        </div>
      </form>
    </Form>
  );
};

export default FamilyForm;
