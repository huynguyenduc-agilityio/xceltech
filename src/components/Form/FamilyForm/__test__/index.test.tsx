import { fireEvent, render, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

import { MutationType } from '@/types';

// Components
import FamilyForm from '..';

const renderComponent = () => {
  const queryClient = new QueryClient();

  return render(
    <QueryClientProvider client={queryClient}>
      <FamilyForm mode={MutationType.Edit} />,
    </QueryClientProvider>,
  );
};

describe('FamilyForm Component', () => {
  it('should match snapshot', () => {
    const container = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it('handle input value change and form submit', () => {
    const { getByPlaceholderText, getByRole } = renderComponent();
    const fullNameInput = getByPlaceholderText('Enter your full name');
    const relationshipInput = getByPlaceholderText('Enter your relationship');
    const phoneInput = getByPlaceholderText('Enter your phone number');
    const addressInput = getByPlaceholderText('Enter your address');
    const submitButton = getByRole('button', { name: 'Update' });

    fireEvent.change(fullNameInput, { target: { value: 'John Doe' } });
    fireEvent.change(relationshipInput, { target: { value: 'Brother' } });
    fireEvent.change(phoneInput, { target: { value: '1234567890' } });
    fireEvent.change(addressInput, { target: { value: '123 Main St' } });

    fireEvent.click(submitButton);

    expect(fullNameInput).toHaveValue('John Doe');
    expect(relationshipInput).toHaveValue('Brother');
    expect(phoneInput).toHaveValue('1234567890');
    expect(addressInput).toHaveValue('123 Main St');
  });

  it('should show error message fields', async () => {
    const { getByPlaceholderText, getByText } = renderComponent();
    const fullNameInput = getByPlaceholderText('Enter your full name');
    const relationshipInput = getByPlaceholderText('Enter your relationship');
    const phoneInput = getByPlaceholderText('Enter your phone number');
    const addressInput = getByPlaceholderText('Enter your address');

    fireEvent.change(fullNameInput, { target: { value: '' } });
    fireEvent.change(relationshipInput, { target: { value: '' } });
    fireEvent.change(phoneInput, { target: { value: '' } });
    fireEvent.change(addressInput, { target: { value: '' } });

    fireEvent.blur(fullNameInput);
    fireEvent.blur(relationshipInput);
    fireEvent.blur(phoneInput);
    fireEvent.blur(addressInput);

    await waitFor(() => {
      expect(getByText('Full Name is required!')).toBeInTheDocument();
      expect(getByText('Relationship is required!')).toBeInTheDocument();
      expect(getByText('Phone No is required!')).toBeInTheDocument();
      expect(getByText('Address is required!')).toBeInTheDocument();
    });
  });
});
