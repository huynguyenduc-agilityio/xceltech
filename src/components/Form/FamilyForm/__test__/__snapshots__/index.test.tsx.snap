// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FamilyForm Component should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="w-full text-black-default space-y-10"
        >
          <div
            class="grid w-full gap-2"
          >
            <div
              class="w-full space-y-5"
            >
              <label
                class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                id="Full Name-label"
              >
                Full Name
              </label>
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                aria-labelledby="Full Name-label"
                autocomplete="off"
                class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                data-testid="input-element"
                id=":r0:-form-item"
                name="fullName"
                placeholder="Enter your full name"
                value=""
              />
            </div>
          </div>
          <div
            class="flex items-start gap-11 justify-between"
          >
            <div
              class="grid w-full gap-2"
            >
              <div
                class="w-full space-y-5"
              >
                <label
                  class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                  id="Relationship-label"
                >
                  Relationship
                </label>
                <input
                  aria-describedby=":r1:-form-item-description"
                  aria-invalid="false"
                  aria-labelledby="Relationship-label"
                  autocomplete="off"
                  class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                  data-testid="input-element"
                  id=":r1:-form-item"
                  name="relationship"
                  placeholder="Enter your relationship"
                  value=""
                />
              </div>
            </div>
            <div
              class="grid w-full gap-2"
            >
              <div
                class="w-full space-y-5"
              >
                <label
                  class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                  id="Phone No-label"
                >
                  Phone No
                </label>
                <input
                  aria-describedby=":r2:-form-item-description"
                  aria-invalid="false"
                  aria-labelledby="Phone No-label"
                  autocomplete="off"
                  class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                  data-testid="input-element"
                  id=":r2:-form-item"
                  name="phone"
                  placeholder="Enter your phone number"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="space-y-2 w-full"
          >
            <label
              class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
            >
              Address
            </label>
            <textarea
              aria-describedby=":r3:-form-item-description"
              aria-invalid="false"
              class="flex min-h-[80px] w-full rounded-regular bg-blue-light py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none resize-none disabled:cursor-not-allowed disabled:opacity-50 px-11"
              id=":r3:-form-item"
              name="address"
              placeholder="Enter your address"
            />
          </div>
          <div
            class="w-full h-[70px]"
          >
            <button
              class="font-productsans inline-flex items-center justify-center whitespace-nowrap rounded-[14px] text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed text-white disabled:hover:bg-primary h-[70px] py-6 w-[364px] bg-green-primary"
              disabled=""
              type="submit"
            >
              Update
            </button>
          </div>
        </div>
      </form>
      ,
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="w-full text-black-default space-y-10"
      >
        <div
          class="grid w-full gap-2"
        >
          <div
            class="w-full space-y-5"
          >
            <label
              class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
              id="Full Name-label"
            >
              Full Name
            </label>
            <input
              aria-describedby=":r0:-form-item-description"
              aria-invalid="false"
              aria-labelledby="Full Name-label"
              autocomplete="off"
              class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
              data-testid="input-element"
              id=":r0:-form-item"
              name="fullName"
              placeholder="Enter your full name"
              value=""
            />
          </div>
        </div>
        <div
          class="flex items-start gap-11 justify-between"
        >
          <div
            class="grid w-full gap-2"
          >
            <div
              class="w-full space-y-5"
            >
              <label
                class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                id="Relationship-label"
              >
                Relationship
              </label>
              <input
                aria-describedby=":r1:-form-item-description"
                aria-invalid="false"
                aria-labelledby="Relationship-label"
                autocomplete="off"
                class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                data-testid="input-element"
                id=":r1:-form-item"
                name="relationship"
                placeholder="Enter your relationship"
                value=""
              />
            </div>
          </div>
          <div
            class="grid w-full gap-2"
          >
            <div
              class="w-full space-y-5"
            >
              <label
                class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                id="Phone No-label"
              >
                Phone No
              </label>
              <input
                aria-describedby=":r2:-form-item-description"
                aria-invalid="false"
                aria-labelledby="Phone No-label"
                autocomplete="off"
                class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                data-testid="input-element"
                id=":r2:-form-item"
                name="phone"
                placeholder="Enter your phone number"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="space-y-2 w-full"
        >
          <label
            class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
          >
            Address
          </label>
          <textarea
            aria-describedby=":r3:-form-item-description"
            aria-invalid="false"
            class="flex min-h-[80px] w-full rounded-regular bg-blue-light py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none resize-none disabled:cursor-not-allowed disabled:opacity-50 px-11"
            id=":r3:-form-item"
            name="address"
            placeholder="Enter your address"
          />
        </div>
        <div
          class="w-full h-[70px]"
        >
          <button
            class="font-productsans inline-flex items-center justify-center whitespace-nowrap rounded-[14px] text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed text-white disabled:hover:bg-primary h-[70px] py-6 w-[364px] bg-green-primary"
            disabled=""
            type="submit"
          >
            Update
          </button>
        </div>
      </div>
    </form>
    ,
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
